pyqtgraph/tests/image_testing.py

```diff
@@ -0,0 +1,520 @@
+else:
+    import httplib
+    import urllib
+from ..Qt import QtGui, QtCore
+from .. import functions as fn
+from .. import GraphicsLayoutWidget
+from .. import ImageItem, TextItem
+
+
+# This tag marks the test-data commit that this version of pyqtgraph should
+# be tested against. When adding or changing test images, create
+# and push a new tag and update this variable.
+testDataTag = 'test-data-3'
+
+
+tester = None
+
+
+def getTester():
+    global tester
+    if tester is None:
+        tester = ImageTester()
+    return tester
+
+
+def assertImageApproved(image, standardFile, message=None, **kwargs):
+    """Check that an image test result matches a pre-approved standard.
+
+    If the result does not match, then the user can optionally invoke a GUI
+    to compare the images and decide whether to fail the test or save the new
+    image as the standard.
+
+    This function will automatically clone the test-data repository into
+    ~/.pyqtgraph/test-data. However, it is up to the user to ensure this repository
+    is kept up to date and to commit/push new images after they are saved.
+
+    Run the test with the environment variable PYQTGRAPH_AUDIT=1 to bring up
+    the auditing GUI.
+
+    Parameters
+    ----------
+    image : (h, w, 4) ndarray
+    standardFile : str
+        The name of the approved test image to check against. This file name
+        is relative to the root of the pyqtgraph test-data repository and will
+        be automatically fetched.
+    message : str
+        A string description of the image. It is recommended to describe
+        specific features that an auditor should look for when deciding whether
+        to fail a test.
+
+    Extra keyword arguments are used to set the thresholds for automatic image
+    comparison (see ``assertImageMatch()``).
+    """
+    if isinstance(image, QtGui.QWidget):
+        w = image
+        image = np.zeros((w.height(), w.width(), 4), dtype=np.ubyte)
+        qimg = fn.makeQImage(image, alpha=True, copy=False, transpose=False)
+        painter = QtGui.QPainter(qimg)
+        w.render(painter)
+        painter.end()
+
+    if message is None:
+        code = inspect.currentframe().f_back.f_code
+        message = "%s::%s" % (code.co_filename, code.co_name)
+
+    # Make sure we have a test data repo available, possibly invoking git
+    dataPath = getTestDataRepo()
+
+    # Read the standard image if it exists
+    stdFileName = os.path.join(dataPath, standardFile + '.png')
+    if not os.path.isfile(stdFileName):
+        stdImage = None
+    else:
+        pxm = QtGui.QPixmap()
+        pxm.load(stdFileName)
+        stdImage = fn.imageToArray(pxm.toImage(), copy=True, transpose=False)
+
+    # If the test image does not match, then we go to audit if requested.
+    try:
+        if image.shape != stdImage.shape:
+            # Allow im1 to be an integer multiple larger than im2 to account
+            # for high-resolution displays
```

Comment:
If im1 can be larger than im2, is it also possible for im2 to be larger than im1 if it was saved on a high res display?